framework/include/utils/ValueCache.h

```diff
@@ -121,3 +127,29 @@ ValueCache<T>::guess(const std::vector<Real> & in_val, T & out_val, Real & dista
   out_val = _data[return_index];
   return true;
 }
+
+template <typename T>
+std::vector<std::tuple<std::vector<Real> &, T &, Real>>
+ValueCache<T>::kNearestNeighbors(const std::vector<Real> & in_val, const std::size_t k)
+{
+  std::vector<std::tuple<std::vector<Real> &, T &, Real>> nearest_neighbors;
+
+  nanoflann::KNNResultSet<Real> result_set(std::min(k, _data.size()));
+  std::vector<std::size_t> return_indices(std::min(k, _data.size()));
+  std::vector<Real> distances(std::min(k, _data.size()));
+
+  result_set.init(return_indices.data(), distances.data());
+
+  // kNN search
+  _kd_tree->findNeighbors(result_set, in_val.data());
+
+  if (result_set.size() < k)
+    mooseWarning(
+        "Requested ", k, " nearest neighbors but current cache size is ", result_set.size());
```

Comment:
If you emit a warning here you should add a method to query the cache `size()`.

Reply:
Done. 
I've also moved this warning earlier. Since the `result_set` size is set as `min(_data.size(), k)`, there is no need to search & check based on `result_set.size()`